### PR TITLE
MDEV-24366 Add a S3 storage engine test case for Huawei OBS

### PR DESCRIPTION
**Pull request by zhaorenhai**
If we use Huawei OBS for S3 storage engine, the amazon test case
will be skiiped, so here we need add a test case for Huawei OBS.

### CODE REVIEW

**Comment by grooverdan** [comment]
I'd suggest as an alternate strategy so we don't end up including every S3 provider.

Like ./plugin/auth_gssapi/mysql-test/auth_gssapi/suite.pm, use mysql-test/suite/s3/suite.pm to set environment variables S3_{HOST,PORT,PROTOCOL,HTTP,BUCKET,ACCESS_KEY,SECRET_KEY} if there is no environment variables set. These should use the defaults listed in mysql-test/suite/s3/my.cnf.

Then get mysql-test/suite/s3/my.cnf to be based off the environment variables (like @ENV.HA_S3_SO is currently).

This way the defaults are preserved but easily set-able by a user running a S3 test against any provider. It also means the full S3 test suite can be applied to Huawei OBS.

**Comment by grooverdan** [comment]
It addition to changing the base of the pull request. Rebase your local commit and push to the same branch. This will update the PR.

**Comment by ottok** [comment]
I removed myself from the reviewers. This is not my domain.

File: mysql-test/suite/s3/hw_obs.test [hunk 1]
```diff
@@ -0,0 +1,23 @@
+--source include/have_s3.inc
```
Comment (vuvova) [inline mysql-test/suite/s3/hw_obs.test:1]: is this test still needed? Does it test some huaweicloud specific functionality?

Comment (zhaorenhai) [reply mysql-test/suite/s3/hw_obs.test:1]: Hello, vuvova, I just simulated amazon test case.

Comment (grooverdan) [reply mysql-test/suite/s3/hw_obs.test:1]: The s3/amazon.test is there to test the s3_protocol_version="Original". From the version in hw_obs test this protocol isn't supported in huaweicloud.

With the other changes you have made, huawei obs can be checked with the appropriate set of ENV variables and this explictit test isn't required.

File: mysql-test/suite/s3/suite.pm [hunk 2]
```diff
@@ -4,5 +4,45 @@ package My::Suite::S3;
...
+}
+
+if (!$ENV{'S3_BUCKET'})
+{
+  return "Environment variable S3_BUCKET need to be set";
+}
+
+if (!$ENV{'S3_REGION'})
+{
+  return "Environment variable S3_REGION need to be set";
```
Comment (grooverdan) [inline mysql-test/suite/s3/suite.pm:19]: https://mariadb.com/kb/en/s3-storage-engine-system-variables/#s3_region this defaults to blank. At least the MinIO can handle a blank region.  Can Huawei handle a blank region?

So just set an empty env variable here.

File: mysql-test/suite/s3/suite.pm [hunk 3]
```diff
@@ -4,5 +4,45 @@ package My::Suite::S3;
...
+}
+
+if (!$ENV{'S3_PROTOCOL_VERSION'})
+{
+  $ENV{'S3_PROTOCOL_VERSION'} = "Auto";
+}
+
+if (!$ENV{'S3_PORT'})
+{
+  $ENV{'S3_PORT'} = 80;
```
Comment (grooverdan) [inline mysql-test/suite/s3/suite.pm:39]: I've just updated the documentation, this default is `0` and its probably better if this does the same, less variables to set.

File: mysql-test/suite/s3/suite.pm [hunk 4]
```diff
@@ -4,5 +4,45 @@ package My::Suite::S3;
...
+}
+
+if (!$ENV{'S3_USE_HTTP'})
+{
+  $ENV{'S3_USE_HTTP'} = "ON";
```
Comment (grooverdan) [inline mysql-test/suite/s3/suite.pm:44]: The default of this is OFF. Probably better to leave this as being the default.

File: mysql-test/suite/s3/suite.pm [hunk 5]
```diff
@@ -4,5 +4,45 @@ package My::Suite::S3;
 
 return "Need S3 engine" unless $::mysqld_variables{'s3'} eq "ON" or $ENV{HA_S3_SO};
 
+if (!$ENV{'S3_HOST_NAME'})
+{
```
Comment (grooverdan) [inline mysql-test/suite/s3/suite.pm:8]: Before returning failure here, we've got our existing CI test suite using the MinIO so merging this code as is will fail our test suite.

So in the code where this github comment is, can you please:
* include client tcp connection to localhost:9000 (https://www.networkworld.com/article/2778731/connection-testing-with-perl.html)
* if that tcp works set the env vars to the values previously in mysql-test/suite/s3/my.cnf so the existing test runs don't need to change (additionally `S3_USE_HTTP=ON`))
* if tcp fails, return the text below.

**Review by grooverdan — CHANGES_REQUESTED** [review]
Please remove the hw_obs.{test,result}. This can be tested with explicit environment variables set.

Thanks for the s3/suite.pm code. Can I just get the additional code so our existing test suite doesn't fail.

**Comment by grooverdan** [comment]
Thank you @zhaorenhai for making all those changes.

ee69c153d9a20675dcfb368ab52b7a63d15f112d is merged into 10.5

**Pull request closed without merging**